proposals/consolidate-merged-lifecycle-hook-schema.md

```diff
@@ -0,0 +1,142 @@
+# Goal
+
+This proposal aims to raise a concept of the 'mergedConfiguration' into the specification, so that implementing tools can generate, as well as consume, a 'merged' `devcontainer.json` file that is the computed merge after parsing a user 'devcontainer.json'.  [The merging logic is already documented](https://containers.dev/implementors/spec/#merge-logic). This proposal focuses on standardizing an output JSON format to represent this merge.
+
+Properties that can not or should not be represented as `devcontainer.json` properties are prepended with a `$`.  This proposal standardizes how these properties should be processed and outputted to file, so that implementing tools can write/consume them consistently.
+
+## Motivation
+
+The current generated 'mergedConfiguration' returned by the `read-configuration` CLI command does not return a `devcontainer.json`. [(visualization)](https://github.com/devcontainers/cli/pull/390#issuecomment-1430190326).  By standardizing the output format, implementing tools can generate a portable, merged `devcontainer.json` that can be consumed by other tools.
+
+## Proposal
+
+### A. Update the devcontainer.json schema
+
+#### A1. Extends lifecycle script properties
+
+Add a value (`LifecycleCommandParallel[]`) to the unioned definition of each lifecycle hook (except 'initializeCommand').  This allows for a syntax that captures the existing allowed types for a lifecycle hook and additionally provides a pattern for attaching additional metadata to a merged configuration.
+
+```typescript
+LifecycleCommand = string | string[]
+LifecycleCommandParallel = { [key: string]: LifecycleCommand; $origin?: string };
+```
+
+In a `devcontainer.json`, each of the following lifecycle hooks shown below can now be represented by one of these three types.
+
+```typescript
+{
+    // ...other devcontainer.json properties...
+	onCreateCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+	updateContentCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+	postCreateCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+	postStartCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+	postAttachCommand?: LifecycleCommand | LifecycleCommandParallel | LifecycleCommandParallel[]
+}
+```
+
+An optional parameter `$origin` is added to the `LifecycleCommandParallel` that supporting tools can use to indicate the source of the command.  For example, this is useful for outputting in the creation log which Feature provided a certain lifecycle hook.  The `$` notation is used to indicate this property is additional tooling metadata that should not be present in a user `devcontainer.json`.
```

Comment:
What is the `$origin` of a bit of config that comes from the `devcontainer.metadata` image label (assuming it doesn't carry one already)?

Reply:
If a section of metadata in `devcontainer.metadata` doesn't already have an `$origin`, we wouldn't anything.  It had to come from somewhere (either a base dev container that defined the image, a Feature, or the user's current dev container).  If omitted, we don't have any insight into where it came from and we'd need to log something generically in those cases.